src/vw/src/main/scala/VowpalWabbitBase.scala

```diff
@@ -0,0 +1,393 @@
+  setDefault(additionalFeatures -> Array())
+  def getAdditionalFeatures: Array[String] = $(additionalFeatures)
+  def setAdditionalFeatures(value: Array[String]): this.type = set(additionalFeatures, value)
+
+  // to support Grid search we need to replicate the parameters here...
+  val numPasses = new IntParam(this, "numPasses", "Number of passes over the data")
+  setDefault(numPasses -> 1)
+  def getNumPasses: Int = $(numPasses)
+  def setNumPasses(value: Int): this.type = set(numPasses, value)
+
+  val learningRate = new DoubleParam(this, "learningRate", "Learning rate")
+  def getLearningRate: Double = $(learningRate)
+  def setLearningRate(value: Double): this.type = set(learningRate, value)
+
+  val powerT = new DoubleParam(this, "powerT", "t power value")
+  def getPowerT: Double = $(powerT)
+  def setPowerT(value: Double): this.type = set(powerT, value)
+
+  val l1 = new DoubleParam(this, "l1", "l_1 lambda")
+  def getL1: Double = $(l1)
+  def setL1(value: Double): this.type = set(l1, value)
+
+  val l2 = new DoubleParam(this, "l2", "l_2 lambda")
+  def getL2: Double = $(l2)
+  def setL2(value: Double): this.type = set(l2, value)
+
+  val interactions = new StringArrayParam(this, "interactions", "Interaction terms as specified by -q")
+  def getInteractions: Array[String] = $(interactions)
+  def setInteractions(value: Array[String]): this.type = set(interactions, value)
+
+  val ignoreNamespaces = new Param[String](this, "ignoreNamespaces", "Namespaces to be ignored (first letter only)")
+  def getIgnoreNamespaces: String = $(ignoreNamespaces)
+  def setIgnoreNamespaces(value: String): this.type = set(ignoreNamespaces, value)
+
+  val initialModel = new ByteArrayParam(this, "initialModel", "Initial model to start from")
+  def getInitialModel: Array[Byte] = $(initialModel)
+  def setInitialModel(value: Array[Byte]): this.type = set(initialModel, value)
+
+  val cacheRows = new BooleanParam(this, "cacheRows", "Cache rows in multi-pass mode")
+  setDefault(cacheRows -> false)
+  def getCacheRows: Boolean = $(cacheRows)
+  def setCacheRows(value: Boolean): this.type = set(cacheRows, value)
+
+  val enableCacheFile = new BooleanParam(this, "enableCacheFile", "Enable local cache file")
+  setDefault(enableCacheFile -> false)
+  def getEnableCacheFile: Boolean = $(enableCacheFile)
+  def setEnableCacheFile(value: Boolean): this.type = set(enableCacheFile, value)
+
+  // abstract methods that implementors need to provide (mixed in through Classifier,...)
+  def getLabelCol: String
+  def getFeaturesCol: String
+
+  implicit class ParamStringBuilder(sb: StringBuilder) {
+    def appendParam[T](param: Param[T], option: String): StringBuilder = {
+      if (!get(param).isEmpty) {
+
+        param match {
+          case _: StringArrayParam => {
+            for (q <- get(param).get)
+              sb.append(' ').append(option).append(' ').append(q)
```

Comment:
also you can use <array>.mkString(<separator>) here:
```
val prefix = ' ' + option + ' '
prefix + get(param).get.mkString(prefix)
```

Reply:
at least the way I did it it got even more complicated (e.g. handling empty params)